bin/cylc

```diff
@@ -14,643 +15,236 @@
+    local PATH_STR="$1"
+    local ITEM_STR="$2"
+    if [[ -z "${PATH_STR:-}" ]]; then
+        echo "${ITEM_STR}"
+    elif [[ "${PATH_STR}" != "${ITEM_STR}" \
+            && "${PATH_STR}" != "${ITEM_STR}":* ]]; then
+        while [[ "${PATH_STR}" == *:"${ITEM_STR}" ]]; do
+            PATH_STR=${PATH_STR%:$ITEM_STR}
+        done
+        while [[ "${PATH_STR}" == *:"${ITEM_STR}":* ]]; do
+            local PATH_HEAD="${PATH_STR%:$ITEM_STR:*}"
+            local PATH_TAIL="${PATH_STR##*:$ITEM_STR:}"
+            PATH_STR="${PATH_HEAD}:${PATH_TAIL}"
+        done
+        echo "${ITEM_STR}:${PATH_STR}"
+    else
+        echo "${PATH_STR}"
+    fi
+}
+
+abort_bad_command() {
+    echo "${CYLC_NS}: ${UTIL}: unknown utility. Abort." >&2
+    echo "Type \"${CYLC_NS} help all\" for a list of utilities." >&2
+    exit 1
+}
+
+abort_ambiguous_command() {
+    echo "\"${UTIL}\" is too ambiguous to be a valid Cylc command abbreviation."
+    echo "These commands match the abbreviation \"${UTIL}\":"
+    sed -r "s,${CYLC_HOME_BIN}/cylc-,," <<<"${commands_match}"
+#    ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null \
+#       | sed -r "s,${CYLC_HOME_BIN}/cylc-,,"
+    exit 1
+}
+
+run_cylc_command_matched() {
+#    COMMAND=$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"*)
+#    if [[ ! -f "${COMMAND}" || ! -x "${COMMAND}" ]]; then
+#        abort_bad_command
+#    fi
+    CYLC_UTIL="$(sed -r "s,${CYLC_HOME_BIN}/cylc-,," <<<"${commands_match}")"
+    export CYLC_UTIL
+    exec "${commands_match}" "$@"
+}
+
+run_cylc_command() {
+#    COMMAND="${commands_match}"
+#    if [[ ! -f "${COMMAND}" || ! -x "${COMMAND}" ]]; then
+#        abort_bad_command
+#    fi
+    CYLC_UTIL="${UTIL}"
+    export CYLC_UTIL
+    exec "${CYLC_HOME_BIN}/cylc-${UTIL}" "$@"
+}
+
+init_cylc
+
+UTIL="help"
+if (( "$#" > 0 )); then
+    UTIL="$1"
+    shift 1
+fi
+# For speed, assume a correct command has been provided first of all. 
+if [[ -f "${CYLC_HOME_BIN}/cylc-${UTIL}" && "${UTIL}" != 'help'  ]]; then
+    run_cylc_command "$@" 
+fi
+
+# Now process help and command matching.
+
+# Check for help or version options.
+case "${UTIL}" in
+# Deal with categories
+control|information|all|task|license|GPL|admin|preparation|hook|discovery\
+        |utility|util|prep|con|info)
+    if (( "$#" == 0 )); then
+        help_util "${UTIL}"
+    elif (( "$#" > 0 )); then
+        UTIL="$1"
+        # Discard the category qualifier argument by shifting
+        # the arguments along by 1.
+        shift 1
+    fi
+    :;;
+help|h|?|--help|-h)
+    help_util "$@"
+    exit 0
+    :;;
+version|--version|-V)
+    print_version "$@"
+    exit 0
+    :;;
+categories)
+    category_help_command "$@"
+    exit 0
+    :;;
+esac
+
+# User has not given a help, category, or version option.
+# So now deal with matching commands and disambiguation.
+
+# Store any matching file glob as a number for later use
+# Minimises re-running ls
+commands_match="$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null)"
```

Comment:
If you add ` || true` at the end, the command will not cause an error. This will allow you to add `set -eu` back to the header of the script.

commands_match="$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null || true)"